pkg/controllers/machinesetsync/machineset_sync_controller_test.go

```diff
@@ -156,56 +155,280 @@ var _ = Describe("With a running MachineSetSync controller", func() {
+
+		Context("when the CAPI MachineSet does not exist", func() {
+			BeforeEach(func() {
+				By("Creating the CAPI and MAPI MachineSets")
+				mapiMachineSet = mapiMachineSetBuilder.Build()
+				Expect(k8sClient.Create(ctx, mapiMachineSet)).Should(Succeed())
+
+				By("Setting the AuthoritativeAPI to ClusterAPI")
+				Eventually(k.UpdateStatus(mapiMachineSet, func() {
+					mapiMachineSet.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityClusterAPI
+				})).Should(Succeed())
+			})
+
+			It("should update the synchronized condition on the MAPI MachineSet", func() {
+				Eventually(k.Object(mapiMachineSet), timeout).Should(
+					SatisfyAll(
+						HaveField("Status.Conditions", ContainElement(
+							SatisfyAll(
+								HaveField("Type", Equal(consts.SynchronizedCondition)),
+								HaveField("Status", Equal(corev1.ConditionFalse)),
+								HaveField("Severity", Equal(machinev1beta1.ConditionSeverityError)),
+								HaveField("Reason", Equal("FailedToGetCAPIMachineSet")),
+								HaveField("Message", Equal("error synchronizing CAPI to MAPI, CAPI machineset not found")),
+							))),
+					))
+
+			})
+
 		})
+
+		Context("When the MAPI MachineAuthority is set to Migrating", func() {
+			BeforeEach(func() {
+				By("Creating the CAPI and MAPI MachineSets")
+				mapiMachineSet = mapiMachineSetBuilder.Build()
+				capiMachineSet = capiMachineSetBuilder.Build()
+
+				Expect(k8sClient.Create(ctx, capiMachineSet)).Should(Succeed())
+				Expect(k8sClient.Create(ctx, mapiMachineSet)).Should(Succeed())
+
+				By("Setting the AuthoritativeAPI to Migrating")
+				Eventually(k.UpdateStatus(mapiMachineSet, func() {
+					mapiMachineSet.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityMigrating
+				})).Should(Succeed())
+			})
+
+			It("should not make any changes", func() {
+				resourceVersion := mapiMachineSet.GetResourceVersion()
+				Consistently(k.Object(mapiMachineSet), timeout).Should(
+					HaveField("ResourceVersion", Equal(resourceVersion)),
+				)
+			})
+		})
+
+		Context("When the MAPI MachineAuthority is not set", func() {
+			BeforeEach(func() {
+				By("Creating the CAPI and MAPI MachineSets")
+				mapiMachineSet = mapiMachineSetBuilder.Build()
+				capiMachineSet = capiMachineSetBuilder.Build()
+
+				Expect(k8sClient.Create(ctx, capiMachineSet)).Should(Succeed())
+				Expect(k8sClient.Create(ctx, mapiMachineSet)).Should(Succeed())
+
+				By("Setting the AuthoritativeAPI to Migrating")
+				Eventually(k.UpdateStatus(mapiMachineSet, func() {
+					mapiMachineSet.Status.AuthoritativeAPI = ""
+				})).Should(Succeed())
+			})
+
+			// I think there may be another case here, where we have had it previously set - synchronized and then it gets unset.
+			// We may want to ensure that we don't change resourceVersion, and that the condition does not change?
+			// Changes to false with a failure reason?
+			//
+			// I'm not sure if we have an API validation that stops this.
+			// TODO: clarify / work out
```

Comment:
I believe we have validation to say that it cannot be cleared once populated, but lets check that

Reply:
I don't think we have a validation currently, here's where I've looked: 

https://github.com/openshift/api/blob/master/machine/v1beta1/types_machineset.go#L158-L167 

looks like we've got one that forces moving through Migrating on the status. 

https://github.com/openshift/api/blob/master/machine/v1beta1/types_machineset.go#L64-L77 

on the spec we say it must be one of CAPI or MAPI. 

I'm not sure I see anything stopping a particularly determined user from manually editing the status? Or am I missing something

Reply:
Perhaps we should add a validation to make sure the field cannot be removed once set 🤔

Reply:
So adding something like:
```
// +kubebuilder:validation:XValidation:rule="self != '' || oldSelf == ''",message="Once authoritativeAPI is set, it cannot be unset."
```
to the API?

Reply:
I see so setting this at the to of `MachineSetStatus`
and:

```
// +kubebuilder:validation:XValidation:rule="self.authoritativeAPI != '' || oldSelf.authoritativeAPI == ''",message="The authoritativeAPI field cannot be unset once it has been set."
```
?

And if this is the right approach, will this also pass validation in case of creation without authoritativeAPI set?

Reply:
Yes, something like that should work. Since it includes `oldSelf`, it's considered a transition rule, and therefore doesn't execute on create. So a status without a valid value on create will be accepted still, which is what we need